### PR TITLE
commit_templater: make various attributes accessible to extensions

### PR DESCRIPTION
**Pull request by torquestomp**
Ideally I also want to make the caching functionality extensible here, but I'm not sure of a good way to go about that yet.  I really want some sort of thread-safe `HashMap<TypeId, Box<dyn Any (of TypeId)>>` that can hold arbitrary cache objects in it but I don't know if that exists.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Review by yuja — APPROVED** [review]

**Comment by yuja** [comment]
> Ideally I also want to make the caching functionality extensible here, but I'm not sure of a good way to go about that yet.

If the cache doesn't have to be shared between methods, it could be baked into a property closure. (e.g. `format` in `"format" of builtin_timestamp_methods()` is a cache.) The `TemplateBuildMethodFn` could also be changed to `Box<dyn Fn` to allow capturing, but that might get uglier.

**Comment by torquestomp** [comment]
> > Ideally I also want to make the caching functionality extensible here, but I'm not sure of a good way to go about that yet.
> 
> If the cache doesn't have to be shared between methods, it could be baked into a property closure. (e.g. `format` in `"format" of builtin_timestamp_methods()` is a cache.) The `TemplateBuildMethodFn` could also be changed to `Box<dyn Fn` to allow capturing, but that might get uglier.

In our case it would need to be shared between methods so capturing wouldn't be sufficient.  I'll try rolling something simple to handle this then, it might be useful for revamping the revset extensions PR too.

File: cli/src/commit_templater.rs [hunk 1]
```diff
@@ -121,6 +121,18 @@ impl<'repo> TemplateLanguage<'repo> for CommitTemplateLanguage<'repo> {
 // If we need to add multiple languages that support Commit types, this can be
 // turned into a trait which extends TemplateLanguage.
 impl<'repo> CommitTemplateLanguage<'repo> {
+    pub fn repo(&self) -> &dyn Repo {
```
Comment (yuja) [inline cli/src/commit_templater.rs:124]: maybe `'repo` lifetime is missing here. (didn't notice before, sorry.)

**Comment by yuja** [comment]
> The `TemplateBuildMethodFn` could also be changed to `Box<dyn Fn` to allow capturing, but that might get uglier.

Fyi, the generic templater uses this approach (because it seemed useful for ad-hoc templating.)
https://github.com/martinvonz/jj/pull/3217/commits/f345d996c21e8b71f47812a103ce176fca1505de#diff-4e078e210000c4918cb327fe441cf124ab9f8b76a3ab9b825209575792a482e4R143

**Pull request merged**